digest/src/newtype.rs

```diff
@@ -0,0 +1,254 @@
+use crate::{
+    core_api::BlockSizeUser,
+    crypto_common::hazmat::{DeserializeStateError, SerializableState, SerializedState},
+    CustomizedInit, FixedOutput, FixedOutputReset, HashMarker, Output, OutputSizeUser, Reset,
+    Update,
+};
+use core::{fmt, marker::PhantomData};
+
+// https://github.com/taiki-e/pin-project/issues/102#issuecomment-540472282
+#[doc(hidden)]
+#[derive(Clone, Default)]
+pub struct Wrapper<'a, T> {
+    pub inner: T,
+    _marker: PhantomData<&'a ()>,
+}
+
+impl<T> BlockSizeUser for Wrapper<'_, T>
+where
+    T: BlockSizeUser,
+{
+    type BlockSize = <T as BlockSizeUser>::BlockSize;
+}
+
+impl<T> CustomizedInit for Wrapper<'_, T>
+where
+    T: CustomizedInit,
+{
+    #[inline]
+    fn new_customized(customization: &[u8]) -> Self {
+        Wrapper {
+            inner: T::new_customized(customization),
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<T> fmt::Debug for Wrapper<'_, T>
+where
+    T: fmt::Debug,
+{
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        T::fmt(&self.inner, f)
+    }
+}
+
+impl<T> HashMarker for Wrapper<'_, T> where T: HashMarker {}
+
+impl<T> OutputSizeUser for Wrapper<'_, T>
+where
+    T: OutputSizeUser,
+{
+    type OutputSize = <T as OutputSizeUser>::OutputSize;
+}
+
+impl<T> Update for Wrapper<'_, T>
+where
+    T: Update,
+{
+    fn update(&mut self, data: &[u8]) {
+        self.inner.update(data)
+    }
+}
+
+impl<T> SerializableState for Wrapper<'_, T>
+where
+    T: SerializableState,
+{
+    type SerializedStateSize = <T as SerializableState>::SerializedStateSize;
+
+    // Required methods
+    fn serialize(&self) -> SerializedState<Self> {
+        self.inner.serialize()
+    }
+    fn deserialize(
+        serialized_state: &SerializedState<Self>,
+    ) -> Result<Self, DeserializeStateError> {
+        Ok(Wrapper {
+            inner: T::deserialize(serialized_state)?,
+            _marker: PhantomData,
+        })
+    }
+}
+
+impl<T> FixedOutput for Wrapper<'_, T>
+where
+    T: FixedOutput,
+{
+    #[inline]
+    fn finalize_into(self, out: &mut Output<Self>) {
+        self.inner.finalize_into(out)
+    }
+}
+
+impl<T> Reset for Wrapper<'_, T>
+where
+    T: Reset,
+{
+    #[inline]
+    fn reset(&mut self) {
+        self.inner.reset()
+    }
+}
+
+impl<T> FixedOutputReset for Wrapper<'_, T>
+where
+    T: FixedOutputReset,
+{
+    #[inline]
+    fn finalize_into_reset(&mut self, out: &mut Output<Self>) {
+        self.inner.finalize_into_reset(out)
+    }
+}
+
+#[macro_export]
+macro_rules! newtype {
+    ($doc: expr, $name: ident, $mod_name: ident, $inner: ty) => {
+        #[doc = $doc]
+        pub struct $name(::digest::newtype::Wrapper<'static, $inner>);
+
+        mod $mod_name {
+            use super::*;
+            use ::digest::{
```

Comment:
IIRC you can use `$crate` to import the items from the current crate..

Reply:
good point.